feat: reindex all memories, files and knowledges

mahenning opened the pull request:
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) and describe your changes before submitting a pull request.

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Please verify that the pull request targets the `dev` branch.
- [x] **Description:** Provide a concise description of the changes made in this pull request.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [x] **Documentation:** Have you updated relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs), or other documentation sources?
- [x] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [x] **Testing:** Have you written and run sufficient tests to validate the changes?
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

- This PR adds a button into the Documents tab in admin view to re-index all memories, single files (chats, files in knowledges..) and knowledge bases, and indicates its progress

### Added

Frontend:
- Adds a button to the "Admin->Documents" tab to re-index all memories, files, knowledges that are currently in the database
- Adds the information on how many files/knowledges have to be re-indexed to the confirmation button, also a stop button
- Adds a progress bar to indicate how long the user should wait for it to finish
Backend:
- Adds /reindex endpoints in utils, memory, files and knowledge. The endpoint in /utils starts the reindexing of all 3 other parts and is the preferred starting point
- Adds a /reindex/stop endpoin to halt reindexing manually
- Adds a "/count" endpoint in memories, files, knowledges to get the total number of items without straining the db to retrieve the actual item contents -> fast & memory efficent
- Adds paginated db access for memories, files, knowledges -> prevents that e.g. all files are read simultaneously into RAM
- Adding redis support and non-blocking threading to support multi-instance deployments (**currently not tested -> Feedback is highly appreciated!**)
---

### Additional Information

- This PR builds on https://github.com/open-webui/open-webui/pull/12571, which added a button to re-index all knowledges. But this re-indexes only the "knowledge-as-a-whole", not the single files in the kowledge (or any other files). Here, the functionality is extended to single files, which may be in the chat but also in knowledges, and also to memories.

### Screenshots or Videos

The confirmation window shows how many files and knowledges will be worked on to get an estimation of the time it will take. It will also enable you to set the number of documents that are loaded into RAM at the same time, and if documents shouldbe processed completely from scratch (takes longer)
<img width="665" height="416" alt="Screenshot 2025-09-05 142054" src="https://github.com/user-attachments/assets/cbf19317-77dc-4855-ae40-c6d914a7d04d" />


A loading bar indicates the progress of the re-indexing
<img width="677" height="54" alt="Screenshot 2025-09-08 144039" src="https://github.com/user-attachments/assets/fbc78488-36e6-411e-834f-47891ab32fdd" />
<img width="657" height="59" alt="Screenshot 2025-09-08 144119" src="https://github.com/user-attachments/assets/04b20483-e9fc-4900-9329-1918a6c8c2ba" />



### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [CONTRIBUTOR_LICENSE_AGREEMENT](CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

mahenning committed:
Add files reindexing button

mahenning committed:
Add progress bar to UI

mahenning committed:
Remove test logging

mahenning committed:
Handles audio files

mahenning commented:
I seek feedback for the following:

- "Reindex all Files Vectors" is not a very good description I think, it is very technical. Anyone has another wording? Maybe "Reindexing all single files"? The important part is that it also re-indexes files from knowledges, so if you include a file from a knowledge with "#" in the chat, the "single" file is used there.
- For full re-indexing, memory indexing is missing. The implementation is easy, but where to put the button in the UI? In the documents tab, where the embedding model is set and the other buttons for re-indexing are, or in another part of the UI?

mahenning commented:
Yes. Also when the chunk size/overlap is updated or the text splitter is changed.

jayteaftw commented:
@mahenning Hey nice work!
Quick question:could this be used for transitioning to a different vector database without losing data?
For example, you start off using Chroma and realize that it is too small to handle the user base so you switch to qdrant. 
Once they set env qdrant to be the DB, you could simply press the reindex all button in the Admin page and it would create the collections for you. 

I think it would make sense to do memory within this PR asweel.
However it also looks like there are some other things being indexed too scuh as web_searches and hashes (?)

```python
# backend/open_webui/retrieval/vector/dbs/qdrant_multitenancy.py
# lines 61-65 
self.MEMORY_COLLECTION = f"{self.collection_prefix}_memories"
self.KNOWLEDGE_COLLECTION = f"{self.collection_prefix}_knowledge"
self.FILE_COLLECTION = f"{self.collection_prefix}_files"
self.WEB_SEARCH_COLLECTION = f"{self.collection_prefix}_web-search"
self.HASH_BASED_COLLECTION = f"{self.collection_prefix}_hash-based"
```

Or are the web search and hash based collections just faster lookup and processing time? And if they were deleted, it wouldnt crash open webui and they would simply be regenerated upon request.

mahenning commented:
I did not test database migration, but in theory this should work, yes. It just iterates over every file, deletes it from vector storage (if it's there) and then re-processes it completely. If there is a new vector store, it just re-processes the files into it. I'd be great to get feedback on this, if possible.

I agree on the memory re-indexing but I got no real feedback on this (or if this PR gets merged at all), so it's on hold right now. Personally, I would not add a third button to re-index memory separately, but merge knowledge/files/memory re-indexing into a single button. While this means potentially heavy lifting on the server side, I don't see a use-case where you only want to re-index just a part of your vector db. Either all or nothing. Everything else is puzzling for the user.

I looked up where the get/insert/upsert methods of the vector db are used and it seems not in web search right now, so I think the lines you posted are just a precaution to future uses.

jayteaftw commented:
> I did not test database migration, but in theory this should work, yes. It just iterates over every file, deletes it from vector storage (if it's there) and then re-processes it completely. If there is a new vector store, it just re-processes the files into it. I'd be great to get feedback on this, if possible.
> 
> I agree on the memory re-indexing but I got no real feedback on this (or if this PR gets merged at all), so it's on hold right now. Personally, I would not add a third button to re-index memory separately, but merge knowledge/files/memory re-indexing into a single button. While this means potentially heavy lifting on the server side, I don't see a use-case where you only want to re-index just a part of your vector db. Either all or nothing. Everything else is puzzling for the user.
> 
> I looked up where the get/insert/upsert methods of the vector db are used and it seems not in web search right now, so I think the lines you posted are just a precaution to future uses.

Unless I am missing something, I totally agree! Especially since the embedding engine is global to every category that is we don't have a separate embedding engine for knowledge/files/memory then it doesn't make sense to just reindex knowledge and files when memory will have the previous embedding engine even if you reindex, effectively defeating the point of reindexing. 

Honestly I would add the changes as sometimes it can take a while to get PRs incorporated as there are a good of amount of PRs to be reviewed.

tjbck commented on backend/open_webui/routers/files.py:
```diff
@@ -221,6 +223,75 @@ async def search_files(
+        except Exception as e:
+            log.error(f"Error deleting file 'file-{file.id}' from vector store: {str(e)}")
+        try:
+            if file.meta['content_type'] in [
```
We might want to use a helper function here to ensure this matches with file upload endpoint

mahenning replied:
Yeah this would be a good idea. I could write one, but I'm not sure where. Is there any place for helper functions?

guenhter replied:
Maybe the whole block from line 248-269 can be reused from `files.py>upload_file`. I guess you anyway have this code from there. Maybe in the files.py, the interesting lines can be put to an own function (e.g. `process_file_by_content_type`) and you then just reuse that function.

mahenning commented on backend/open_webui/routers/files.py:
```diff
@@ -221,6 +223,75 @@ async def search_files(
+
+# This function is just used to get a progress bar about the re-indexing to the UI
+@router.get("/reindex/stream")
+async def stream_progress():
```
I did dig a bit through the code and saw that `AppConfig` uses redis (if configured). But it also enforces the values to be saved to be `PersistentConfig`, which is too much for a temporal runtime value I think. It would be nice if I can use redis for this. Is that an option or do you have another idea? I think that UI feedback on this functionality would be important, as it could run for a long time.

guenhter replied:
I guess Redis would be nice, but it not running everywhere...

It is not the prettiest thing, but what about a file `/var/run/lock/foo/owui-file-reindex.lock` where the progress is written to? The reindex endpoint is writing it, and this `stream_progress` function is only reading.

Furthermore, it should then not be possible to trigger a second reindex-operation, if there is already a reindex task running. 

What do you think?

mahenning replied:
Personally I only have experience using a single replica. I see 3 options:

A) Add a warning to the confirmation pop-up that this should only be used in single instances. Easiest to implement, but then it would be a feature that cant be used by everyone -> puzzling, when to use or not? What happens if a user in multi-replica setup presses "go" either way?

B) Use a local .lock file. I'm not sure if this is worth pursuing, this only works for very local multi-replica setups.

C) Use redis (somehow) or another method to enable this feature for multi-replica instances. I thought that redis is a dependency for multi-replica setups, because of [this blue Info box](https://docs.openwebui.com/getting-started/env-configuration/#redis) and also the [requirements for the helm charts](https://github.com/open-webui/helm-charts/tree/main/charts/open-webui#requirements). 

When going C), I can add the logic to use redis if configured, and fall back to the local logic that is implemented now, which works for single replicas.

mahenning committed:
Merge remote-tracking branch 'upstream/dev' into fix--reindex-all-files

mahenning committed:
Merge remote-tracking branch 'upstream/dev' into fix--reindex-all-files

mahenning committed:
Hopefully fixed the multi-replica problem with reindex status.
Added getting the loading bar when re-loading the page.
Blocked the reindex button when already reindexing.

mahenning committed:
Merge remote-tracking branch 'upstream/dev' into fix--reindex-all-files

mahenning commented:
- Added Redis support for reindexing status, fallback to local variable if redis is not configured
- Starting the reindexing and getting the loading bar blocks the reindexing button, so you can't start it twice
- Implemented that the loading bar re-appears if the page is re-loaded or one comes back to the page
- Added a "Counting files.. " dialogue to the start when one clicks the reindex button for responsiveness while actually counting files in the background, which can take a while

In my one-replica setup, this is working now, with full responsiveness of the UI while the reindexing is working. You can now even leave/reload the page and get back, and the loading bar is still there and updated.

! I can not test this with redis/multi-replica settings, because I dont have the multi-replica setup for that. If someone has, and can test that the reindexing status is shared across instances, it would be great. I *think* that it should work in theory.

mahenning committed:
Reduced the size of the API call greatly by omitting files content from counting files.

mahenning committed:
Added memories to reindex process.

mahenning committed:
Make progress bar not disappear after reindexing to indicate finished job.

Add delay to backend stream to not end between different endpoints to reindex.

mahenning committed:
Add knowledges to reindex process.

mahenning committed:
Merge remote-tracking branch 'upstream/dev' into fix--reindex-all-files

mahenning committed:
Move frontend js functions to the correct files.

mahenning committed:
Removed the other reindex button.

mahenning commented:
- The button now reindexes all memories, files and knowledges in this order. The confirmation dialogue shows the number of files and knowledges (screenshot below).
- The loading bar now stays after the progress is finished to better indicate that the progress is done. When reloading the page after reindexing, it will be hidden again.
- I've removed the other button to just reindex the knowledge bases, as it is included and would just be confusing to have an additional button for 1/3 functionality

The helper function TODO is still open but maybe thats work for another PR.
In my opinion this PR is ready to merge. The multi-replica setting is still untested, but I cannot test it.

<img width="562" height="236" alt="grafik" src="https://github.com/user-attachments/assets/94ee4e34-79bc-4ccb-8ce8-8f22dc8a3402" />

<img width="1493" height="91" alt="grafik" src="https://github.com/user-attachments/assets/c75da876-d1a4-44b5-966b-05b345df8780" />

mahenning commented:
Quick info: With chromadb, the size of the vector store almost doubles when reindexing. The problem is that chromadb does not delete the vector- and metadata when a collection is deleted (contrary to their [own documentation](https://docs.trychroma.com/docs/collections/delete-data)) 
> Chroma supports deleting items from a collection by id using .delete. The embeddings, documents, and metadata associated with each item will be deleted.

https://github.com/open-webui/open-webui/pull/16520 fixes the deletion of vector folders, but the internal chroma.sqlite3 database still holds all the metadata of the files, including full text blocks of the (deleted) documents. A workaround is to backup and delete the `vector_db` folder in `backend/data/` and start the re-indexing. This will build a fresh, clean vector store.

Classic298 commented:
>  but the internal chroma.sqlite3 database still holds all the metadata of the files,

absolutely, i noticed that - despite running the VACUUM command on chroma db, which my PR does do.

I will try to identify a way to make it delete even the remaining data.

mahenning committed:
Merge remote-tracking branch 'upstream/dev' into fix--reindex-all-files

mahenning committed:
Add "/count" endpoint for files, memory, knowledges (admin-only)
Add paginated iteration function for files, memory, knowledge to not load everything at once into RAM

mahenning committed:
Update /reindex endpoints to use new count and pagination, also ensures threading.

mahenning committed:
Formatting

mahenning committed:
Formatting leftovers

mahenning committed:
Hopefully final formatting..

mahenning committed:
Merge remote-tracking branch 'upstream/dev' into fix--reindex-all-files

mahenning committed:
Add a new /reindex endpoint in utils to unify reindexing endpoints in the backend.
Change REINDEX_PROGRESS to send progress and current status of all 3 reindex endpoints.
Split progress bar in 3 parts and color each part when endpoint is finished.

elabbarw commented:
> @elabbarw I've updated the code with your snippets (thanks again for them, it made it so much easier for me). If you have time to review, I'd be very glad! I put the compute-heavy parts into async calls, and tested with both CPU/GPU the responsiveness of the app with a ping to /health every 1s.
> 
> Currently, the batch_size is hard-coded to a batch of 10. I did no yet test the difference in performance for e.g. 100/1000, but it should be more or less negligible as it only leads to more paginated db calls, which should be fast in comparison to the workload of embedding. But more user control would be nice, maybe with an env var? Also, for knowledges it reads all files for the current knowledge. This _could_ lead to RAM problems on constrained systems with a very, very large knowledge base. To solve this, it could be also paginated/split:
> 
> * read out all file ids for current knowledge
> * get `batch_size` files from the list of ids, process them, then get the next batch
> 
> Let me know if the env var or pagination for knowledge files should be implemented as well.

- I think it would good to provide the option in admin settings with the default set at 10. Env var is the easy option, hehe 😉 
- I totally agree with batching the files as well! Good call 😄

mahenning committed:
Fixed the embedding of single files form knowledges
Added /reindex/stop endpoint for manually halting the process
Added a utton to stop
Added the option to not read conten from disk, but from db
Added tooltips

mahenning commented:
To give a few insights into the implementation:
I have decided not to use environment variables for batch size and the "read from disk" feature. It will use default vars that are set in the frontend and backend separately. A bit ugly, but it is much work to let the frontend load that from the backend. The impact of a (reasonable) value of batch_size (3, 5, 20..) is negligible. I got the same speed for batch size 3 as with batch size 20. As this only controls how many files are read from the db, and db access is usually fast, 10 shouldbe fine for 99% of cases. The files are processed one after the other either way.

As for the "Process files from disk" feature.. I'm not completly happy with the solution, but thats based on how OWUI handles file reads right now. I saw that reading the file from disk took ~25-40% of the total reindexing time per file. And with the same context extraction parameters, the extracted content is (almost) the same that is already saved in the db.
There are 2 points that could be improved:
- Currently, the file content saved in the db is one long concatenated string. The file content read from disk usually is a list of strings that are split by page, which means that a chunk of text cannot be over multiple pages. Reading the content from the db changes that, as there are no "pages" anymore. To change that, one have to change the way the content is saved in the db, which exceeds this PR
- When reading from disk, there are two steps: 
a) Actually reading the doc and extracting the content (I/O bound)
b) Embedding the content (usually CPU/GPU bound)
My idea was a queue, so that document A can be embedded while document B can already be read from disk at the same time. The problem here is that the `process_file` function does both steps, so I can not split them up into their own processes.

mahenning committed:
Merge remote-tracking branch 'upstream/dev' into fix--reindex-all-files

mahenning committed:
Merge remote-tracking branch 'upstream/dev' into fix--reindex-all-files

i-iooi-i commented:
https://github.com/open-webui/open-webui/pull/18565

Can the quantity of retrieved memories be controlled or adjusted? Integer (1-100), controls max memories retrieved

mahenning commented:
@i-iooi-i maybe I don't understand the question, but this PR is exclusively for re-indexing the vector data for memory (files, knowledges..). It does not touch how memories works for the user.

i-iooi-i commented:
@mahenning My apologies, it seems I was mistaken. The current functional experience with OpenWebUI is quite unsatisfactory. Despite storing numerous memories, it seems capable of recalling only a handful, or sometimes none at all. I hope the official team will soon consider bolstering this particular feature. In the interim, I've resorted to using a plugin as a replacement for the memory function; it performs admirably on the new OpenWebUI v0.6.41 and boasts excellent memory retrieval capabilities.

https://www.reddit.com/r/OpenWebUI/comments/1o24giv/comment/nqirhjk/
<img width="1080" height="477" alt="image" src="https://github.com/user-attachments/assets/3f17ccd4-0b66-47d9-ac98-9fc8fbb56d58" />

jayteaftw commented:
Any update on this?

stronk7 commented:
It would so nice to get ALL contents (memories, knowledges, files, ...) reindexed each time ANY change in embeddings (model, vectors db, ... ) happens. Right now I'm blocked using an old embedding when, right now, there are greatly improved ones to be used. And, also, I want to move away from built-in chroma to a remote one... 

And this PR will solve, if I'm not wrong, all those situations.

Thanks for the effort and contributions, @mahenning, much appreciated. Hope this finally lands and we get all those (embedding-changes) use cases covered. They are real.

Ciao :-)

Classic298 commented:
@TomTheWise

No or yes depending on your Vector Database.

The PR deletes each collection via VECTOR_DB_CLIENT.delete_collection() before re-embedding, so it depends on what your vector DB backend does with that call. If delete_collection fully drops the collection/table (like Milvus Multitenancy does), then yes, the new embeddings with different dimensions will work because the schema gets recreated from scratch on insert.

If your pgvector implementation only truncates rows but keeps the table structure (and its dimension constraint), then no, you'd still hit a dimension mismatch. You'd want to verify that your specific vector DB adapter actually drops the underlying table/index, not just the data.

stevenmoto commented:
Trying to change the embedding model in my OWUI instance.

For existing files, the reindex button, causes an error:
chromadb.errors.InvalidArgumentError: Collection expecting embedding with dimension of 2560, got 2048

I believe this work would help to process them correctly?

madsgraaew commented:
> Trying to change the embedding model in my OWUI instance.
> 
> For existing files, the reindex button, causes an error: chromadb.errors.InvalidArgumentError: Collection expecting embedding with dimension of 2560, got 2048
> 
> I believe this work would help to process them correctly?

No, apparently they dont delete or overwrite the existing collection, so unless your new embedding model has the same dimensions, the reindex button does nothing.
